### PR TITLE
ovn-kubernetes: enable TLS

### PR DESCRIPTION
**Pull request by squeed**
This creates a certificate authority and a single certificate. It configures all the OVN processes to use that certificate:
- nbdb and sbdb serve behind that certificate
- nbdb and sbdb authenticate clients against the CA
- ovn-controller and ovnkube-node use the certificate to authenticate

### CODE REVIEW

File: bindata/network/ovn-kubernetes/ovnkube-node.yaml [hunk 1]
```diff
@@ -184,8 +189,14 @@ spec:
...
-            --nb-address "tcp://${db_ip}:{{.OVN_NB_PORT}}" \
-            --sb-address "tcp://${db_ip}:{{.OVN_SB_PORT}}" \
+            --nb-address "ssl://${db_ip}:{{.OVN_NB_PORT}}" \
+            --nb-client-privkey /ovn-cert/tls.key \
+            --nb-client-cert /ovn-cert/tls.crt \
+            --nb-client-cacert /ovn-ca/ca-bundle.crt \
+            --sb-address "ssl://${db_ip}:{{.OVN_SB_PORT}}" \
+            --sb-client-privkey /ovn-cert/tls.key \
+            --sb-client-cert /ovn-cert/tls.crt \
+            --sb-client-cacert /ovn-ca/ca-bundle.crt \
```
Comment (pecameron) [inline bindata/network/ovn-kubernetes/ovnkube-node.yaml]: @squeed  These should be passed in the ovnkube-config configmap

File: bindata/network/ovn-kubernetes/ovnkube-master.yaml [hunk 2]
```diff
@@ -145,7 +155,9 @@ spec:
               - -c
               - |
                 retries=0
-                while ! ovn-sbctl -t 5 set-connection ptcp:{{.OVN_SB_PORT}} -- set connection . inactivity_probe=0; do
+                while ! ( ovn-sbctl set-ssl /ovn-cert/tls.key /ovn-cert/tls.crt /ovn-ca/ca-bundle.crt && \
+                          ovn-sbctl -t 5 set-connection pssl:{{.OVN_SB_PORT}} -- set connection . inactivity_probe=0 )
+                do
```
Comment (pecameron) [inline bindata/network/ovn-kubernetes/ovnkube-master.yaml]: @squeed This hack only works at install time. Setting the listen port was done after the sdb was synced. Once set the port is opened immediately at sdb start, it doesn't wait for the resync to complete.

File: bindata/network/ovn-kubernetes/ovnkube-node.yaml (hunk 1 above)
Comment (squeed) [reply bindata/network/ovn-kubernetes/ovnkube-node.yaml]: That needs to pass CI first.

**Comment by squeed** [comment]
All the tests that failed are either expected or seem like uninteresting flakes. Nevertheless,

/test e2e-aws-ovn-kubernetes

File: bindata/network/ovn-kubernetes/ovnkube-master.yaml (hunk 2 above)
Comment (squeed) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: Hmm, not sure what you mean by this. How should this line read, then?

Comment (danwinship) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: The whole `postStart` thing is really insane anyway... there must be a nicer way to do this

File: bindata/network/ovn-kubernetes/ovnkube-master.yaml [hunk 3]
```diff
@@ -133,7 +143,7 @@ spec:
           fi
           exec /usr/share/openvswitch/scripts/ovn-ctl run_sb_ovsdb \
             --no-monitor \
-            --ovn-nb-log="-vconsole:${OVN_LOG_LEVEL} -vfile:off"
+            --ovn-sb-log="-vconsole:${OVN_LOG_LEVEL} -vfile:off"
```
Comment (danwinship) [inline bindata/network/ovn-kubernetes/ovnkube-master.yaml]: Can you make this a separate commit? (eg, if we ended up having to revert the SSL stuff there's no reason to revert this too)

File: bindata/network/ovn-kubernetes/ovnkube-master.yaml (hunk 2 above)
Comment (pecameron) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: @squeed @dcbw 
The install uses the unix socket to set up and sync the sdb. When it is ready, the ports are opened. On successive reboots of sdb the port is immediately opened and it starts serving requests. It doesn't wait for sync to complete (this is what prevented install from working when the ports were opened early). There needs to be a lock that prevents this.

ovnkube should set the ports when the DB is set up, it should not be in the script. At present there is a race since ovnkube is working in parallel with the script. This appears to work because on install (first boot) there is not much to. It shouldn't be done in the script.

The "while" is not needed since ovnkube has been working with the sdb over the unix socket. The first attempt will work.

Comment (squeed) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: I see that it configures the endpoints in ovnkube-master, but where does it wait for northd to do a sync loop? And why does this matter? What happens if the southd is listening as configuration is enabled?

Comment (pecameron) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: sdb immediately serves requests when it starts. When I originally did the ovn bringup this caused pods to go into crashloopbackoff. I had to hold off sdb until it was in sync with the api server. So I used the unix socket to set it up and when that was done setup the network ports.

Comment (pecameron) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: The cluster knows what should be running. Bad things happen whenever the sdb has a different  world view than the cluster. sdb it not truth. It is a cache. It needs to be treated like that.

Comment (pecameron) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: @squeed @danwinship  ^^

**Comment by squeed** [comment]
Great, a panic:

```
time="2019-10-18T13:32:48Z" level=info msg="Deleting pod: success"
E1018 13:32:48.654451       1 runtime.go:69] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:76
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:65
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:51
/usr/local/go/src/runtime/panic.go:522
/usr/local/go/src/runtime/panic.go:82
/usr/local/go/src/runtime/signal_unix.go:390
/usr/local/go/src/sync/mutex.go:182
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/namespace.go:82
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/pods.go:183
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/ovn.go:143
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:209
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:251
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:172
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:54
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:170
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:209
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:556
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:265
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:548
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:546
/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71
/usr/local/go/src/runtime/asm_amd64.s:1337
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x46e97b]

goroutine 70 [running]:
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/runtime/runtime.go:58 +0x105
panic(0x13c2740, 0x24b1d80)
	/usr/local/go/src/runtime/panic.go:522 +0x1b5
sync.(*Mutex).Unlock(0x0)
	/usr/local/go/src/sync/mutex.go:182 +0x2b
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).deletePodFromNamespaceAddressSet(0xc0006be000, 0xc000cbfb70, 0x10, 0xc000c23a20, 0xb)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/namespace.go:82 +0x40c
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).deleteLogicalPort(0xc0006be000, 0xc00098f400)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/pods.go:183 +0x435
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn.(*Controller).WatchPods.func3(0x1540900, 0xc00098f400)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/ovn/ovn.go:143 +0x45
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnDelete(...)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:209
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache.FilteringResourceEventHandler.OnDelete(0xc000a1c0c0, 0x17af280, 0xc000860ea0, 0x1540900, 0xc00098f400)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:251 +0x69
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*WatchFactory).newFederatedHandler.func3.1(0xc0003165a0)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:172 +0x126
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*informer).forEachHandler(0xc0002ffe90, 0x1540900, 0xc00098f400, 0xc000435cf0)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:54 +0x16d
github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory.(*WatchFactory).newFederatedHandler.func3(0x1540900, 0xc00098f400)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/pkg/factory/factory.go:170 +0x20c
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnDelete(...)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/controller.go:209
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1.1(0xc00016cf00, 0xc00017e730, 0x0)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:556 +0x22b
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait.ExponentialBackoff(0x989680, 0x3ff0000000000000, 0x3fb999999999999a, 0x5, 0x0, 0xc000435e38, 0x42b47f, 0xc00017e760)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:265 +0x51
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache.(*processorListener).run.func1()
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:548 +0x79
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1(0xc000055768)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:152 +0x54
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil(0xc000435f68, 0xdf8475800, 0x0, 0x42bd01, 0xc000396780)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:153 +0xf8
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait.Until(...)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:88
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache.(*processorListener).run(0xc00034ca80)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/client-go/tools/cache/shared_informer.go:546 +0x9c
github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1(0xc000330920, 0xc0003928b0)
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x4f
created by github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start
	/go-controller/_output/go/src/github.com/ovn-org/ovn-kubernetes/go-controller/vendor/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x62
```

**Comment by squeed** [comment]
we definitely broke something
/test e2e-aws-ovn-kubernetes
watching CI.

File: bindata/network/ovn-kubernetes/ovnkube-master.yaml (hunk 2 above)
Comment (squeed) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: This is all moot, because ovn-kubernetes communicates over a unix socket to the local DB. So we can't use it's feature that programs the endpoints in to the nb/sbdb.

Comment (squeed) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: Hmm. This is all moot, because ovnkube-master uses a unix socket to talk to the sbdb. It needs an independent way to learn its endpoints.

**Comment by squeed** [comment]
Looks like the ovnk rebase fixed the test issues. This is now green.

**Comment by squeed** [comment]
filed https://github.com/ovn-org/ovn-kubernetes/pull/867 to fix the namespace panic, which is unrelated to this PR.

File: bindata/network/ovn-kubernetes/ovnkube-master.yaml (hunk 2 above)
Comment (pecameron) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: @squeed It uses the unix socket to sync up the sdb. The sdb must be setup before the tcp ports are opened or pods go into crashloopbackoff.  This was done as a hack at install time. Restarts and reboots are still a problem since sdb remembers to immediately open the ports and start serving. 
At present we start master and then set up the tcp ports. The expectation is that on install, master will get the sdb setup before the ports get setup (this is a race). This does not work for restarts/reboots.
All this should be moved to ovnkube --init-master

Comment (squeed) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: Why do pods start crashlooping?

Comment (dcbw) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: @squeed we have code in config.go to do the set-ssl operation already. This doesn't need to be done in the script.

Comment (dcbw) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: Well, unless you want to split the master up and run the sbdb on a different node than master... but this PR doesn't do that. I suppose we should have a conversation about how much the master should handle DB configuration; does the admin/container need to set up the authentication and ports, or ovnkube-master?

Doing it all in the containers would be great and we could delete code from Go. But this PR doesn't do that...

Comment (pecameron) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: @dcbw @squeed we need to use unix sockets while setting up sdb. When sdb listens on the network port it handles requests. This must be prevented while setting up sdb otherwise pods go into CrashLoopBackup. The listening ports need to be set up after DB sync completes.  (This was discovered in original OVN install work).

When sdb comes up it immediately listens on all configured ports and handles requests. This is  a problem whenever the sdb and apiserver are out of sync. The unix socket hack only works during install, not restart or reboot.

Comment (squeed) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: @dcbw 

> we have code in config.go to do the set-ssl operation already.

Reading the code, I couldn't figure out how to get go to set-ssl and also use a unix socket.

Comment (squeed) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: @pecameron 

>  This must be prevented while setting up sdb otherwise pods go into CrashLoopBackup.

As I already explained, I don't care if they go in CrashLoopBackoff, because they quickly leave it. Besides, this was fixed when we stopped getting the chassis id from the sbdb. Also, none of this will matter, since we have to open the sbdb port for raft.

File: bindata/network/ovn-kubernetes/ovnkube-master.yaml (hunk 3 above)
Comment (squeed) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: dine

File: bindata/network/ovn-kubernetes/ovnkube-master.yaml [hunk 4]
```diff
@@ -93,7 +97,9 @@ spec:
               - -c
               - |
                 retries=0
-                while ! ovn-nbctl -t 5 set-connection ptcp:{{.OVN_NB_PORT}} -- set connection . inactivity_probe=0; do
+                while ! ( ovn-nbctl set-ssl /ovn-cert/tls.key /ovn-cert/tls.crt /ovn-ca/ca-bundle.crt && \
```
Comment (dcbw) [inline bindata/network/ovn-kubernetes/ovnkube-master.yaml]: @squeed instead of `set-ssl`, why aren't we passing the SSL options to ovn-ctl in the first place?

```
  --ovn-nb-db-ssl-key=KEY OVN Northbound DB SSL private key file
  --ovn-nb-db-ssl-cert=CERT OVN Northbound DB SSL certificate file
  --ovn-nb-db-ssl-ca-cert=CERT OVN Northbound DB SSL CA certificate file
  --ovn-sb-db-ssl-key=KEY OVN Southbound DB SSL private key file
  --ovn-sb-db-ssl-cert=CERT OVN Southbound DB SSL certificate file
  --ovn-sb-db-ssl-ca-cert=CERT OVN Southbound DB SSL CA certificate file
```

Comment (squeed) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: As if that worked. It doesn't.

File: bindata/network/ovn-kubernetes/ovnkube-master.yaml (hunk 2 above)
Comment (dcbw) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: @pecameron we do have Numans HA pr which won't update the endpoints that the nodes use until the southbound is synced with northbound. So now the way this will work is that ovn-controller won't even know where to look for the southbound until things are already synced, which avoids the southbound port open issue.

File: bindata/network/ovn-kubernetes/ovnkube-master.yaml (hunk 4 above)
Comment (dcbw) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: @squeed not as far as I can tell with `set -x` at the top of the ovn-ctl script. They get passed to the database process.

Comment (squeed) [reply bindata/network/ovn-kubernetes/ovnkube-master.yaml]: This totally works. Duh. Fixed.

**Comment by squeed** [comment]
rebased, tested, everything works.

**Comment by squeed** [comment]
/retest
huh
